cmd/clusterctl/client/tree/discovery.go

```diff
@@ -93,54 +104,131 @@ func Discovery(ctx context.Context, c client.Client, namespace, name string, opt
+	}
+
+	machinePoolList, err := getMachinePoolsInCluster(ctx, c, cluster.Namespace, cluster.Name)
+	if err != nil {
+		return nil, err
 	}
 
-	if len(machinesList.Items) == len(controlPlaneMachines) {
+	if len(machinesList.Items) == len(controlPlaneMachines) && len(machinePoolList.Items) == 0 {
+		// TODO: revise this check when machinepoolmachines is implemented
 		return tree, nil
 	}
 
 	workers := VirtualObject(cluster.Namespace, "WorkerGroup", "Workers")
 	tree.Add(cluster, workers)
 
+	err = addMachineDeploymentToObjectTree(ctx, c, cluster, workers, machinesList, tree, options, addMachineFunc)
+	if err != nil {
+		return nil, err
+	}
+
+	addMachinePoolsToObjectTree(ctx, c, cluster, workers, machinePoolList, tree, options)
+
+	// Handles orphan machines.
+	if len(machineMap) < len(machinesList.Items) {
+		other := VirtualObject(cluster.Namespace, "OtherGroup", "Other")
+		tree.Add(workers, other)
+
+		for i := range machinesList.Items {
+			m := &machinesList.Items[i]
+			if _, ok := machineMap[m.Name]; ok {
+				continue
+			}
+			addMachineFunc(other, m)
+		}
+	}
+
+	return tree, nil
+}
+
+func addAddonsToObjectTree(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, tree *ObjectTree) {
+	if resourceSetBindings, err := getResourceSetBindingsInCluster(ctx, c, cluster.Namespace, cluster.Name); err == nil {
+		// Should this be in it's own group?
+		addons := VirtualObject(cluster.Namespace, "AddonGroup", "Addons")
+		tree.Add(cluster, addons)
+		for i, resourceSetBinding := range resourceSetBindings {
+			tree.Add(addons, &resourceSetBindings[i], ObjectMetaName("ClusterResourceSetBinding"))
+			for _, binding := range resourceSetBinding.Spec.Bindings {
+				if resourceSet, err := getResourceSetInCluster(ctx, c, cluster.Namespace, binding.ClusterResourceSetName); err == nil {
+					tree.Add(addons, resourceSet, ObjectMetaName("ClusterResourceSet"))
+				}
+			}
+		}
+	}
+}
+
+func addControlPlane(ctx context.Context, c client.Client, cluster *clusterv1.Cluster, controlPlane *unstructured.Unstructured, tree *ObjectTree, options DiscoverOptions) {
+	tree.Add(cluster, controlPlane, ObjectMetaName("ControlPlane"), GroupingObject(true))
+
+	if options.ShowAllResources {
+		// Add control plane infrastructure ref using spec fields guaranteed in contract
+		infrastructureRef, found, err := unstructured.NestedMap(controlPlane.UnstructuredContent(), "spec", "machineTemplate", "infrastructureRef")
+		if err != nil && found {
+			infrastructureObjectRef := &corev1.ObjectReference{
+				Kind:       infrastructureRef["kind"].(string),
+				Namespace:  infrastructureRef["namespace"].(string),
+				Name:       infrastructureRef["name"].(string),
+				APIVersion: infrastructureRef["apiVersion"].(string),
+			}
+			if machineTemplate, err := external.Get(ctx, c, infrastructureObjectRef, cluster.Namespace); err == nil {
+				tree.Add(controlPlane, machineTemplate, ObjectMetaName("MachineInfrastructureTemplate"))
```

Comment:
This relates to the initial comment above, what is the goal of loading templates in bulk at T0? have you considered alternatives, like incremental loading?

Reply:
Also, is it required to read the actual template, given that we are not using the info contained in it?

Reply:
For this and the CRS, I thought `tree.Add()` needed a `client.Object` to add a CRD, and so far in the function we've been reading in the object using the object ref. If we don't need to find an additional ref within it and just want to add the kind, name, and namespace, could we add a "dummy" node containing that info?

Reply:
Yes, an unstructured object can do the trick.
It doesn't help in keeping the size of the tree under control, but at least we save API calls to retry info we don't care about

Reply:
Sounds good, I think that would work. The only thing is that by using a placeholder object, we aren't directly checking to verify that the object actually exists before adding it to the tree. Do we want to add a check for that before adding the unstructured object?